compiler/rustc_trait_selection/src/solve/alias_relate.rs

```diff
@@ -84,3 +93,105 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
         }
     }
 }
+
+enum IgnoreAliases {
+    Yes,
+    No,
+}
+
+impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
+    /// In case a rigid term refers to a placeholder which is not referenced by the
+    /// alias, the alias cannot be normalized to that rigid term unless it contains
+    /// either inference variables or these placeholders are referenced in a term
+    /// of a `Projection`-clause in the environment.
+    fn alias_cannot_name_placeholder_in_rigid(
+        &mut self,
+        param_env: ty::ParamEnv<'tcx>,
+        rigid_term: ty::Term<'tcx>,
+        alias: ty::Term<'tcx>,
+    ) -> bool {
+        // Check that the rigid term is actually rigid.
+        if rigid_term.to_alias_term().is_some() || alias.to_alias_term().is_none() {
+            return false;
+        }
+
+        // If the alias has any type or const inference variables,
+        // do not try to apply the fast path as these inference variables
+        // may resolve to something containing placeholders.
+        if alias.has_non_region_infer() {
+            return false;
+        }
+
+        let mut referenced_placeholders =
+            self.collect_placeholders_in_term(rigid_term, IgnoreAliases::Yes);
+        for clause in param_env.caller_bounds() {
```

Comment:
Would be nice to explain in detail what this is searching for. Specifically, we disqualify any of our rigid term's placeholders if they show up in projection goals, since that may assist in normalizing.

e.g. if we have

```
rigid = W<!A>
alias = Alias<!B>
param-env = [Alias<!B> normalizes-to W<!A>, ...]
```

then we exclude `!A` from our rigid placeholder list, right?

I think this may also be worth mentioning that this doesn't really care about how the placeholders show up in the term of the projection predicate, since this is really just a quick heuristic.